### DISCUSSION TITLE
Error with skewness plot

### DISCUSSION THREAD

**Discussion by giannoko** [pyvista/pyvista#3568] [Q&A] [answered]
Hi,

I am trying to plot the skewness of the sample sphere mesh of pyvista and I am getting -1 for all cells. Here is the sample of the code
What is the problem? Thank you

```
	mesh = pv.Sphere(theta_resolution=20, phi_resolution=20)

	mesh = mesh.compute_cell_quality(quality_measure=quality)
	plotter = pv.Plotter(shape=(1, 1),border=True, border_color = "grey", window_size=[1024, 768] )
	plotter.subplot(0, 0)

	actor = plotter.add_mesh_clip_plane(
			mesh,
			origin=(0.,0.,0.),
			normal='z',
                        widget_color="blue",
			implicit=False,
			normal_rotation=True,
			crinkle=False,
			color = "white",
			style="surface",
			scalars="CellQuality",
			show_edges=True,
			tubing=True,
			origin_translation=False,
			outline_translation=False,
			edge_color="black",
			opacity=1.0,
			lighting=False,
			cmap="plasma",
			label='Tube mesh',
			show_scalar_bar=True,
			scalar_bar_args={'title': quality , 'color' : "white"},
			render=True
	)

```

![skewness](https://user-images.githubusercontent.com/92250377/200598379-73a22989-c5d9-4ec6-8fa2-fd4c577ae1a0.PNG)

**Comment by darikg** [comment, marked as answer]
I think skewness is only defined for quadrilateral cells

**Reply by banesullivan** [reply]
👍🏻 

`-1` is the default value for `compute_cell_quality `'s `null_value` parameter. This hints that skew cannot be computed for these cells

**Reply by giannoko** [reply]
Hi. @darikg skewness is also applicable to triangles since it evaluates to what extent the edge and angle of the created mesh deviates from the ideal state (equilateral triangle / square). @banesullivan I know that -1 means that skewness cannot be computed for that kind of cells, but I wonder what kind of cells are supported for skewness in pyvista. I would expect to be able to compute the skewness in already implemented meshes in pyvista like the sphere I posted.

**Reply by banesullivan** [reply]
I'm very unfamiliar with these cell quality metrics, and this filter is entirely coming from VTK (without anything too unique from PyVista). I tried looking for some docs on VTK's `vtkCellQuality` filter and it's `SetQualityMeasureToSkew` method without finding anything too insightful. Considering this is a simpler PyVista filter, I might recommend copying the source code from `compute_cell_quality` and opening an topic on the [VTK discourse](https://discourse.vtk.org/) to see if people there have some insight on the "skew" metric for this mesh.

```py
mesh = ... # vtkPolyData with traingles - use vtkSphereSource
alg = vtk.vtkCellQuality()
alg.SetInputData(mesh)
alg.SetQualityMeasureToSkew()  # <-- what does this evaluate?
alg.SetUndefinedQuality(-1)
alg.Updata()
out = alg.GetOutput()
```

Hope this helps

**Reply by MatthewFlamm** [reply]
I haven't had time to delve into super detail, but it looks like there are several 'skew' based choices: 

See https://vtk.org/doc/nightly/html/classvtkMeshQuality.html#a296d81b67b805f32af3c8305bc1b8c99

The vtkCellQuality can choose based on this Enum or by the methods like pyvista is using.  I wonder if you explicitly use `EQUIANGLE_SKEW` whether it will work.

See here for implementation for triangles https://github.com/Kitware/VTK/blob/352c985b9c73847fed2595e4dc5a4d15c3d671c5/Filters/Verdict/vtkCellQuality.cxx#L174-L211

**Reply by darikg** [reply]
note that that's vtkMeshQuality, not vtkCellQuality which is what pyvista uses, though I have no idea what the difference is, also there's some mismatch between the docs and python bindings (this is vtk 9.1.0):

```python
from vtkmodules.vtkFiltersVerdict import vtkCellQuality, vtkMeshQuality

for alg in (vtkCellQuality, vtkMeshQuality):
    for k in dir(alg):
        if 'Skew' in k:
            print(f'{alg.__name__}.{k}')
```

> vtkCellQuality.SetQualityMeasureToSkew
> vtkMeshQuality.HexSkew
> vtkMeshQuality.QuadSkew
> vtkMeshQuality.SetHexQualityMeasureToSkew
> vtkMeshQuality.SetQuadQualityMeasureToSkew

**Reply by giannoko** [reply]
First of all, I would like to thank all of you for your time and consideration. Your replies to any open issue in this channel are exceptional. 

I will try all of your suggestions and I will come back soon for the feedback.

**Reply by MatthewFlamm** [reply]
> note that that's vtkMeshQuality, not vtkCellQuality which is what pyvista uses,

It looks like `vtkCellQuality` method `vtkSetEnumMacro()` uses an enum from `vtkMeshQuality`

The documentation on this functionality is a bit scattered so some testing would be required.

See https://vtk.org/doc/nightly/html/classvtkCellQuality.html#ae466243ee35d7dce2a566fac22c37926

If there is a mismatch in reported methods, it is likely a version difference.  The vtk documentation I linked is the bleeding edge docs.

**Reply by MatthewFlamm** [reply]
Here is a working snippet.  QualityMeasure `29`, i.e. `EQUIANGLE_SKEW` was added in the last year.  I only tested on vtk versions 9.0.x and 9.2.2 and it works on 9.2.2, but not previously.


```py
import pyvista as pv
import vtk
mesh = pv.Sphere()
alg = vtk.vtkCellQuality()
alg.SetInputData(mesh)
alg.SetQualityMeasure(29)
alg.SetUndefinedQuality(-1)
alg.Update()
out = pv.wrap(alg.GetOutput())
out["CellQuality"]
```

**Reply by banesullivan** [reply]
We may want to modify `compute_cell_quality()` to support integers as the `quality_measure` parameter to enable these options since `EQUIANGLE_SKEW` is not covered in that filter's list of metrics.

**Reply by MatthewFlamm** [reply]
I made https://github.com/pyvista/pyvista/issues/3587.

**Reply by giannoko** [reply]
@MatthewFlamm Thank you for your answer. Could you please tell me how to add the above script into "add_mesh_clip_plane" method since I want to display the cell quality through this. My code is like that now
```
mesh = mesh.compute_cell_quality(quality_measure="skew")
plotter.add_mesh_clip_plane(mesh,scalars="CellQuality")
```

**Reply by banesullivan** [reply]
The `out` object from `out = pv.wrap(alg.GetOutput())` is a PyVista mesh object and can be passed to `add_mesh_clip_plane` as `plotter.add_mesh_clip_plane(out, scalars="CellQuality")`

**Reply by giannoko** [reply]
Thank you again for your reply. It works!
